midend/local_copyprop.cpp

```diff
@@ -205,7 +205,16 @@ IR::AssignmentStatement *DoLocalCopyPropagation::preorder(IR::AssignmentStatemen
 }
 
 IR::AssignmentStatement *DoLocalCopyPropagation::postorder(IR::AssignmentStatement *as) {
-    if (as->left == as->right) {   // FIXME -- need deep equals here?
+    if (as->left->toString() == as->right->toString() &&
```

Comment:
Converting an operation to a string generally just returns the operation.  So if you have an assignment like `a[0] = b[1]`, both `toString()` calls will return `"[]"` and this will incorrectly eliminate the assignment.

The right fix (as noted in the FIXME) is to use a deep equality comparison.  Doing that is probably best done by adding a `deep_equals` method in the IR generator that will automatically create a deep comparison operation that could be used here -- on my TODO list, but keeps getting pushed back by other things.